test/e2e/azure.go

```diff
@@ -0,0 +1,223 @@
+		Name:      KUBE_SYSTEM_SECRET_NAME,
+		Namespace: "kube-system",
+	}
+	if err = a.kubeClient.Get(context.Background(), secretName, secret); err != nil {
+		return fmt.Errorf("failed to get credentials secret %s, error : %v", secretName.Name, err)
+	}
+
+	a.config = &cluserConfig{
+		TenantID:                    string(secret.Data[TENANT_ID]),
+		SubscriptionID:              string(secret.Data[SUBSCIPTION_ID]),
+		ResourceGroup:               string(secret.Data[RESOURCE_GROUP]),
+		Location:                    string(secret.Data[REGION]),
+		ClientID:                    string(secret.Data[CLIENT_ID]),
+		ClientSecret:                string(secret.Data[CLIENT_SECRET]),
+		UseManagedIdentityExtension: true,
+	}
+
+	var environment azure.Environment
+	if a.config.Cloud == "" {
+		environment = azure.PublicCloud
+	}
+	a.config.Environment = environment
+
+	return nil
+}
+
+func (a *azureTestHelper) prepareZoneClient() error {
+	token, err := getAccessToken(a.config)
+	if err != nil {
+		return err
+	}
+	a.zoneClient = dns.NewZonesClientWithBaseURI(a.config.Environment.ResourceManagerEndpoint, a.config.SubscriptionID)
+	a.zoneClient.Authorizer = autorest.NewBearerAuthorizer(token)
+	return nil
+}
+
+// Credentials should in as json string for azure provider.
+func (a *azureTestHelper) makeCredentialsSecret(namespace string) *corev1.Secret {
+	credData := struct {
+		TenantID       string `json:"tenantId"`
+		SubscriptionID string `json:"subscriptionId"`
+		ResourceGroup  string `json:"resourceGroup"`
+		ClientID       string `json:"aadClientId"`
+		ClientSecret   string `json:"aadClientSecret"`
+	}{
+		TenantID:       a.config.TenantID,
+		SubscriptionID: a.config.SubscriptionID,
+		ResourceGroup:  a.config.ResourceGroup,
+		ClientID:       a.config.ClientID,
+		ClientSecret:   a.config.ClientSecret,
+	}
+	azureCreds, err := json.Marshal(credData)
+	if err != nil {
+
+	}
+
+	return &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("azure-config-file-%s", randomString(16)),
+			Namespace: namespace,
+		},
+		Data: map[string][]byte{
+			"azure.json": azureCreds,
+		},
+	}
+}
+
+func (a *azureTestHelper) platform() string {
+	return string(configv1.AzurePlatformType)
+}
+
+func (a *azureTestHelper) ensureHostedZone(rootDomain string) (string, []string, error) {
+	location := "global"
+	z, err := a.zoneClient.CreateOrUpdate(context.TODO(), a.config.ResourceGroup, rootDomain,
+		dns.Zone{Location: &location}, "", "")
+	if err != nil {
+		return "", []string{}, err
+	}
+	var zoneID string
+	zoneID = *z.ID
+	nameservers := append(*z.ZoneProperties.NameServers)
+	a.zoneName = rootDomain
+	return zoneID, nameservers, nil
+}
+
+func (a *azureTestHelper) deleteHostedZone(rootDomain string) error {
+	if a.zoneName == "" {
+		fmt.Printf("ZoneName is empty, nothing to be deleted")
+		return nil
+	}
+	if _, err := a.zoneClient.Delete(context.TODO(), a.config.ResourceGroup, a.zoneName, ""); err != nil {
```

Comment:
Why setting `zoneName` to `rootDomain` in `ensure*` method just to use it here when you have `rootDomain` as argument of the method?

Reply:
again it is based on the provider,
aws expects the zoneID, and azure expects zonename.
in operator_test, we are passing zoneId. 

there is refactoring required, i am working on it, once this PR closed i will push the code.

Reply:
Indeed, the zone ID is passed, the name of the argument of the method is wrong then, it should be something like `id` or `zoneID`.
Let's keep this intra-method context of `zoneName` and refactor in the follow ups.